src/main/java/com/bpodgursky/jbool_expressions/NExpression.java

```diff
@@ -17,6 +18,8 @@ protected NExpression(List<? extends Expression<K>> expressions){
     }
 
     this.expressions = expressions.toArray(ExprUtil.<K>expr(0));
+    this.hash = expressions.stream().mapToInt(Expression::hashCode)
```

Comment:
the problem here is that we throw out half of the hash bits by using .hashCode() rather than .hash

pure multiplication does not seem like a good hash code, as any time we get a LSB of 0 it will propagate through.  I guess for all expressions with more than 150 children this hash function returns 0 every time.

so as much as I like streams, I would try copying the java string hash code here:

hash = hashBase; // passed in as a parameter, e.g. "and".hashCode()

for (int i = 0; i < length; i++)
{
   hash *= 31;
   hash += expressions[i].hash
}

Reply:
fixed.